docs/rfcs/027-pageserver-disaster-recovery.md

```diff
@@ -0,0 +1,157 @@
+historic lsn through support staff, downtimes are more affordable, and
+inevitable if the original had been subject to the corruption. Ideally, the
+two code paths would share code, so the solution would be designed for not
+requiring downtimes.
+
+### API endpoints
+
+The proposed design of the new pageserver API endpoint is:
+
+```
+/v1/tenant/:tenant_id/timeline/:timeline_id/recover?from_lsn=<from_lsn>&to_lsn=<lsn>&target=<target_timeline_id>
+```
+
+The `recover` endpoint answers to `POST` requests and issues a WAL recovery
+process. It takes the WAL of the specified timeline, and writes its results
+into a *different* timeline, the target timeline. This two timeline scheme
+ensures that we can stil serve requests and preserves immutability guarantees
+assumed by the [generation numbers RFC], as in corruption situations we might
+indeed end up with different content than the original timeline we are storing
+data from.
+
+Where the parameters are:
+
+* `:tenant_id`: the id of the tenant to recover
+* `:timeline_id`: the id of the timeline to recover
+* `from_lsn`: The lsn at the start of our recovery process.
+  Optional, if not specified the recovery process is fully WAL based.
+  If specified, the recovery process starts from data at the *main* timeline
+  at the specific lsn (maybe we still trust that lsn). In any case, the target
+  timeline must be less advanced than `from_lsn`.
+* `to_lsn`: the end of the lsn range we want to recover. As usual, it's an
+  exclusive limit. This allows playing back the WAL to a specific moment in the
+  past, if this is wanted/needed. E.g. there might be a data-corrupting
+  pageserver crash at lsn 1234 and we don't want that so we play back until
+  1233.
+* `target`: the timeline id we are writing into. The parameter is optional;
+  if no target timeline is provided, a new timeline is created.
+  One specifies either both `from_lsn` and `target`, or neither.
+  Writing into the same timeline is forbidden.
+
+[generation numbers RFC]: https://github.com/neondatabase/neon/pull/4919
+
+### Higher level features
+
+The API endpoint described above allows for a large set of higher level
+features to be implemented.
+
+* recovery not just one timeline, but of multiple timelines. Oncluding, if
+  wanted, all timelines of the tenant.
+* recovery just until a specific moment in time, even if pageserver history is
+  completely turned off.
+* a repeated job to *detect* errors/corruption, both to verify that the backup
+  works and to verify that the backed up original is not corrupted.
+
+## Non Goals
+
+At the danger of being repetitive, the main goal of this feature is to be a
+backup method, so reliability is very important. This implies that other
+aspects like performance or space reduction are less important.
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+Most changes would happen to the pageservers.
+For the higher level features, maybe other components like the console would
```

Comment:
I think the main thing to define about console's involvement is how it can make use of the recovered timelines.

e.g. we might call directly to a pageserver to initiate a recovery to some timeline ID, but what happens when the recovery is done and the user wants to start an endpoint targeting this new timeline ID?   Somehow the console needs to learn about this timeline ID that was created out of band (or, the console needs to pick the ID and initiate the recovery).